scripts/dmsdk/gen_cpp.py

```diff
@@ -37,54 +37,101 @@ def rename_symbols(renames, line):
-    #     return None
+def get_indent(indent):
+    return '    ' * indent
+
+def _find_doc(docs, name):
+    for doc in docs['items']:
+        if doc.get('name', '') == name:
+            return doc
+    return None
+
+def _add_line(lines, tag, s):
+    if s is None:
+        return
+    if not tag:
+        lines.append(s)
+    else:
+        lines.append(f'@{tag} {s}')
+
+def _add_param(lines, tag, doc):
+    if doc is None:
+        return
+    name = doc.get('name', '')
+    typ = doc.get('type', '')
+    desc = doc.get('desc', '')
+    if not typ:
+        lines.append(f'@{tag} {name} {desc}')
+    else:
+        lines.append(f'@{tag} {name} [type:{typ}] {desc}')
+
+def _format_doc_lines(lines, indent):
+    indent_str = get_indent(indent)
+    if isinstance(lines, str):
+        lines = lines.splitlines()
+    return ('\n' + indent_str + ' * ').join(lines)
+
+def gen_doc(docs, name, renamed, indent=1, language='C++'):
+    doc = _find_doc(docs, name)
+    if not doc:
+        return f"""    /*#
+        * Generated from [ref:{name}]
+        */"""
 
     lines = []
-    lines.extend(doc.desc.strip().split('\n'))
-    for param in doc.params:
-        if param.name:
-            lines.append('@%s %s %s' % (param.param, param.name, param.text)) # @param text
-        else:
-            lines.append('@%s %s' % (param.param, param.text)) # @name text
+    _add_line(lines, None, doc.get('brief', ''))
 
-    lines = [l.strip() for l in lines] # fix some annoying white spaces at the beginning of some lines
+    desc = doc.get('desc', doc.get('brief', ''))
+    desc = _format_doc_lines(desc, indent)
+    _add_line(lines, None, desc)
 
-    lines = [rename_symbols(renames, l) for l in lines]
+    doctype = doc.get('type', None)
+    if doctype:
+        lines.append(f'@{doctype}')
+    else:
+        doctype = 'function'
 
-    for example in doc.examples:
-        example = example.desc.strip()
-        if not example.startswith("```cpp"):
-            continue
-        lines.append("@examples")
-        lines.extend(example.split('\n'))
+    _add_line(lines, 'name', renamed)
+    _add_line(lines, 'language', language)
+
+    if doctype in ['function', 'typedef']:
+        for param in doc.get('params', []):
+            _add_param(lines, 'param', param)
+
+        _add_param(lines, 'return', doc.get('return', None))
+
+    elif doctype in ['struct', 'enum']:
+        for member in doc.get('members', []):
+            _add_param(lines, 'member', member)
 
-    if not lines:
-        return '    // no documentation generated'
+    for (_, v) in doc.get('notes', []):
+        v = _format_doc_lines(v, indent)
+        _add_line(lines, 'note', v)
 
-    lines = [lines[0]] + ['     ' + l for l in lines[1:]]
-    return '    /*%s\n' % lines[0] + '\n'.join(lines[1:]) + '\n     */'
+    for (_, v) in doc.get('examples', []):
+        # TODO: Filter out languages that aren't C++
+        if 'objective-c' in v:
+            v = v.replace('\\@', '@')
```

Comment:
There are issues with our regex's that don't support "@" within an example... still not fixed...